File: src/lexer.cc
#include <token.hpp>
#include <lexer.hpp>
#include <vari.hpp>

#include <iostream>

extern Vari g_vari;

bool Lexer::isAtEnd(){ return current >= source.size(); }

char Lexer::peek(){
   if (isAtEnd()) return '\0';
   return source[current];
}

char Lexer::peekNext(){
    if(current+1 >= source.size()) return '\0';
    return source[current+1];
}


bool Lexer::isDigit(char c){ return c >= '0' && c <= '9'; }

bool Lexer::isAlpha(char c){
    return (c >= 'a' && c <= 'z') || 
        (c >= 'A' && c <= 'Z') || c == '_';
}

bool Lexer::isAlphaNumeric(char c){ return isDigit(c) || isAlpha(c); }


char Lexer::next(){ 
    return source[current++]; 
}

// only conumes if current char is specifically c
bool Lexer::match(char c){
    if(isAtEnd()) return false;

    if(source[current] != c) return false;

    current++;
    return true;
}


void Lexer::consumeString(){
    while(peek() != '"' && !isAtEnd()){
        if(peek() == '\n') line++;
        next();
    }

    if(isAtEnd()){
        g_vari.error(line, "Nonterminated string.");
        return;
    }

    next(); // consumes closing quote

    string strval = source.substr(start+1, (current-start) - 1);

    cstr cstrval = {.ptr = strval.c_str(), .len=strval.size()};
    addToken(STRING, {.str = {.ptr=strval.c_str(), .len=strval.size()}});
}

void Lexer::consumeNumber(){
    while(isDigit(peek())) next();

    if(peek() == '.' && isDigit(peekNext())){
        next();

        while(isDigit(peek())) next();
    }

    addToken(NUMBER, {.fp64= std::stod(source.substr(start, current-start)) });
}

Lexer::Lexer(string src){
    start = 0; current = 0; line = 1; 
    source = src;

    keywords.insert({"and", AND});
    keywords.insert({"else", ELSE});
    keywords.insert({"false", FALSE});
    keywords.insert({"for", FOR});
    keywords.insert({"fun", FUN});
    keywords.insert({"if", IF});
    keywords.insert({"nil", NIL});
    keywords.insert({"or", OR});
    keywords.insert({"print", PRINT});
    keywords.insert({"return", RETURN});
    keywords.insert({"this", THIS});
    keywords.insert({"true", TRUE});
    keywords.insert({"let", LET});
    keywords.insert({"while", WHILE});
}

Lexer::~Lexer(){}


void Lexer::addToken(TokenType type,  Object literal){
    string text = source.substr(start, current-start);
    tokens.push_back(Token(type, text, literal, line));
}

void Lexer::addToken(TokenType type){ addToken(type, {.ptr=0}); }

void Lexer::scanToken(){
    char c = next();
    switch(c){
        case '(': addToken(LEFT_PAREN); break;
        case ')': addToken(RIGHT_PAREN); break;
        case '{': addToken(LEFT_BRACE); break;
        case '}': addToken(RIGHT_BRACE); break;
        case ',': addToken(COMMA); break;
        case '.': addToken(DOT); break;
        case '-': addToken(MINUS); break;
        case '+': addToken(PLUS); break;
        case ';': addToken(SEMICOLON); break;
        case '*': addToken(STAR); break;

        // other operators
        case '!': addToken(match('=') ? NE : NOT); break;
        case '=': addToken(match('=') ? DOUBLE_EQ : EQ); break;
        case '<': addToken(match('=') ? LEQ : LT); break;
        case '>': addToken(match('=') ? GEQ : GT); break;
        case '#':
            while(peek() != '\n') next(); 
            break;
        case '/' : addToken(SLASH); break;

        // other "pointless" chars
        case ' ':
        case '\r':
        case '\t':
            break;

        case  '\n': line++; break;
        case '"': consumeString(); break;

        default:
          if(isDigit(c)) 
              consumeNumber();
          else 
              g_vari.error(line, "Unexpected character.");
          break;
    }
}

vector<Token> Lexer::scanTokens(){ 
    while(!isAtEnd()){
        start = current;
        scanToken();
    }

    Object literal {.ptr=0};
    
    Token token(_EOF, "", literal,  line);
    tokens.push_back(token);

    return tokens;
}

File: src/lexer.rs
use std::{any::Any, collections::HashMap, sync::Arc};

use crate::token::{Token, TokenType};
use crate::vari::VARI;

pub struct Lexer {
    keywords: HashMap<String, TokenType>,
    tokens: Vec<Token>,
    source: String,
    line: usize,
    current: usize,
    start: usize,
}

impl Lexer {
    pub fn new(src: String) -> Self {
        let mut keywords_map = HashMap::new();
        keywords_map.insert("and".to_owned(), TokenType::AND);
        keywords_map.insert("else".to_owned(), TokenType::ELSE);
        keywords_map.insert("false".to_owned(), TokenType::FALSE);
        keywords_map.insert("for".to_owned(), TokenType::FOR);
        keywords_map.insert("fun".to_owned(), TokenType::FUN);
        keywords_map.insert("if".to_owned(), TokenType::IF);
        keywords_map.insert("nil".to_owned(), TokenType::NIL);
        keywords_map.insert("or".to_owned(), TokenType::OR);
        keywords_map.insert("print".to_owned(), TokenType::PRINT);
        keywords_map.insert("return".to_owned(), TokenType::RETURN);
        keywords_map.insert("super".to_owned(), TokenType::SUPER);
        keywords_map.insert("this".to_owned(), TokenType::THIS);
        keywords_map.insert("true".to_owned(), TokenType::TRUE);
        keywords_map.insert("var".to_owned(), TokenType::VAR);
        keywords_map.insert("while".to_owned(), TokenType::WHILE);

        Self {
            keywords: keywords_map,
            tokens: vec![],
            source: src,
            line: 1,
            current: 0,
            start: 0,
        }
    }

    fn done(&self) -> bool {
        self.current >= self.source.len()
    }

    fn advance(&mut self) -> char {
        self.current += 1;
        self.source.as_bytes()[self.current - 1] as char
    }

    fn add_token_with_literal(&mut self, tk_type: TokenType, literal: Option<Arc<dyn Any>>) {
        let strval = self.source[self.start..self.current].to_owned();
        self.tokens
            .push(Token::new(tk_type, strval, self.line, literal));
    }

    fn add_token(&mut self, tk_type: TokenType) -> () {
        self.add_token_with_literal(tk_type, None);
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            return '\0';
        }

        self.source.as_bytes()[self.current + 1] as char
    }

    fn peek(&self) -> char {
        if self.done() {
            return '\0';
        }

        self.source.as_bytes()[self.current] as char
    }

    fn is_digit(&self, c: char) -> bool {
        c >= '0' && c <= '9'
    }
    fn is_alpha(&self, c: char) -> bool {
        (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
    }

    fn is_alphanumeric(&self, c: char) -> bool {
        self.is_alpha(c) || self.is_digit(c)
    }

    fn match_expected(&mut self, expected: char) -> bool {
        if self.done() || self.source.as_bytes()[self.current] != expected as u8 {
            return false;
        }

        self.current += 1;
        true
    }

    fn consume_num_literal(&mut self) {
        while self.is_digit(self.peek()) {
            self.advance();
        }

        if self.peek() == '.' && self.is_digit(self.peek_next()) {
            self.advance();

            while self.is_digit(self.peek()) {
                self.advance();
            }
        }

        let num_substr = self.source[self.start..self.current].to_owned();

        match num_substr.parse::<f64>() {
            Ok(val) => {
                self.add_token_with_literal(TokenType::NUMBER, Some(Arc::new(val)));
            }
            _ => {
                VARI.error(self.line, "Invalid number type");
            }
        }
    }

    fn consume_string_literal(&mut self) {
        while self.peek() != '"' && !self.done() {
            if self.peek() == '\n' {
                self.line += 1;
            }

            self.advance();
        }

        // closing "
        self.advance();

        let token_literal_value: String = self.source[self.start + 1..self.current - 1].to_owned();
        self.add_token_with_literal(TokenType::STRING, Some(Arc::new(token_literal_value)));
    }

    fn consume_identifier(&mut self) {
        while self.is_alphanumeric(self.peek()) {
            self.advance();
        }

        // check if IDENTIFIER is a reserved keyword
        let ident = self.source[self.start..self.current].to_owned();
        match self.keywords.get(&ident) {
            Some(keyword) => {
                self.add_token(keyword.to_owned());
            }
            None => {
                self.add_token(TokenType::IDENTIFIER);
            }
        }
    }

    fn scan_token(&mut self) -> () {
        match self.advance() {
            '(' => self.add_token(TokenType::LPAREN),
            ')' => self.add_token(TokenType::RPAREN),
            '{' => self.add_token(TokenType::LBRACE),
            '}' => self.add_token(TokenType::RBRACE),
            ',' => self.add_token(TokenType::COMMA),
            '.' => self.add_token(TokenType::DOT),
            '*' => self.add_token(TokenType::STAR),
            '/' => self.add_token(TokenType::SLASH),
            '+' => self.add_token(TokenType::PLUS),
            '-' => self.add_token(TokenType::MINUS),
            ';' => self.add_token(TokenType::SEMICOLON),

            // ignore whitespace
            ' ' => {}
            '\r' => {}
            '\t' => {}
            '\n' => self.line += 1,

            // operators
            '=' => {
                if self.match_expected('=') {
                    self.add_token(TokenType::ISEQ);
                } else {
                    self.add_token(TokenType::EQUAL);
                }
            }
            '!' => {
                if self.match_expected('=') {
                    self.add_token(TokenType::NE);
                } else {
                    self.add_token(TokenType::NOT);
                }
            }
            '<' => {
                if self.match_expected('=') {
                    self.add_token(TokenType::LE);
                } else {
                    self.add_token(TokenType::LT);
                }
            }
            '>' => {
                if self.match_expected('=') {
                    self.add_token(TokenType::GE);
                } else {
                    self.add_token(TokenType::GT);
                }
            }

            // comments
            '#' => {
                while self.peek() != '\n' && !self.done() {
                    self.advance();
                }
            }
            // string literals
            '"' => {
                self.consume_string_literal();
            }
            c => {
                if self.is_digit(c) {
                    self.consume_num_literal();
                } else if self.is_alpha(c) {
                    self.consume_identifier();
                } else {
                    VARI.error(self.line, format!("Unexpected character").as_str());
                }
            }
        }
    }

    pub fn scan_tokens(&mut self) -> Vec<Token> {
        while !self.done() {
            self.start = self.current;
            self.scan_token();
        }

        self.tokens.push(Token {
            token_type: TokenType::EOF,
            lexeme: "".to_owned(),
            line: self.line,
            literal: None,
        });

        self.tokens.clone()
    }
}
